[BUG] `AttentionLSTM` - Remove activity regularizers

marcosfelt opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->


#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->
Fixes #6685
Also relevant to #6153 (second issue)

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
I fixed the following issues with `AttentionLSTM` used in `LSTMFCNNetwork`:

-  Removed the unused `activity_regularizer`, which was causing errors
-  Changed the name of the first argument in `call` from `inputs` to `sequences` to match what is expected by keras. Apparently, the argument names need to match the arguments from the parent class (RNN, in this case).
- Added `register_keras_serializable` to the code as required by keras (see [here](https://keras.io/guides/serialization_and_saving/#custom-objects))
- Changed `use_chrono_initialization` to `unit_forget_bias` as the former was never initialized, and, therefore, could not be loaded upon deserialization


In case it's helpful, here's the error that pointed me to changing the arguments from `inputs` to `sequences`.
```shell
ValueError: Exception encountered when calling AttentionLSTM.call().

  For a `compute_output_shape()` method with more than one argument, all arguments should have a `_shape` suffix and match an argument from `call()`. E.g. `compute_output_shape(self, foo_shape, bar_shape)`  For layer 'AttentionLSTM', received `compute_output_shape()` argument `sequences_shape`, but `call()` does not have argument `sequences`.
```
#### Does your contribution introduce a new dependency? If yes, which one?
N/A
<!--
Only relevant if you changed pyproject.toml.
We try to minimize dependencies in the core dependency set. There
are also further specific instructions to follow for soft dependencies.
See here for handling dependencies in sktime: https://www.sktime.net/en/latest/developer_guide/dependencies.html
-->

#### What should a reviewer concentrate their feedback on?
- Should this be considered a breaking change due to the change in the arguments?
- Am I correct that there is already sufficient test coverage or do we need to specifically test the saving/loading functionality?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?

<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

If I understand correctly, there are already tests that were failing previously, but here is the code I was using to test the changes:

```python
from sktime.regression.deep_learning import LSTMFCNRegressor
from sktime.datasets import load_italy_power_demand
from sktime.base import load
from datetime import datetime

X_train, y_train = load_italy_power_demand(split="train", return_type="numpy3D")
X_new, _ = load_italy_power_demand(split="test", return_type="numpy3D")
estimator=LSTMFCNRegressor(verbose=0,n_epochs=5,attention = True)
estimator.fit(X_train,y_train)
file_name = datetime.now().strftime("%Y%m%d%H%M%S")
estimator.save(file_name)
new_estimator=load(file_name)

new_estimator.predict(X_new)
```

#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

<!--
Thanks for contributing!
-->

marcosfelt commented:
Happy new year! Did I need to do anything else for this PR to be reviewed?

fkiraly reviewed (changes requested):
Thanks! I need more context to review this PR, would appreciate your explanation.

Can you kindly explain what errors the parameter was causing under which circumstances? It depends on that whether this is a breaking change imo.

marcosfelt commented:
Hi @fkiraly - I've tried to explain some of the context in more detail below:

1.  The parameters 'activity_regularizer' was not able to be set at instantiation of `AttentionLSTMCell`. I think this is due to `activity_regularizer` not being one of the regularizers available in keras.
  <details>
  <summary>Error trace</summary>
   ```bash
   Traceback (most recent call last):
  File "/Users/Kobi/code/sktime/test.py", line 9, in <module>
    estimator.fit(X_train,y_train)
  File "/Users/Kobi/code/sktime/sktime/regression/base.py", line 256, in fit
    self._fit(X, y)
  File "/Users/Kobi/code/sktime/sktime/regression/deep_learning/lstmfcn.py", line 178, in _fit
    self.model_ = self.build_model(self.input_shape)
  File "/Users/Kobi/code/sktime/sktime/regression/deep_learning/lstmfcn.py", line 136, in build_model
    input_layers, output_layer = self._network.build_network(input_shape, **kwargs)
  File "/Users/Kobi/code/sktime/sktime/networks/lstmfcn.py", line 88, in build_network
    x = AttentionLSTM(self.lstm_size)(x)
  File "/Users/Kobi/code/sktime/sktime/networks/lstmfcn_layers.py", line 684, in __init__
    super().__init__(
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/layers/rnn/rnn.py", line 200, in __init__
    super().__init__(**kwargs)
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/layers/layer.py", line 269, in __init__
    self.activity_regularizer = regularizers.get(activity_regularizer)
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/layers/layer.py", line 1443, in __setattr__
    return super().__setattr__(name, value)
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/backend/tensorflow/trackable.py", line 28, in __setattr__
    super().__setattr__(name, value)
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/tensorflow/python/trackable/autotrackable.py", line 70, in __setattr__
    super(AutoTrackable, self).__setattr__(name, value)
  AttributeError: can't set attribute 'activity_regularizer'
  ````
  </details>
  
2. Another issue during instantiation of `AttentionLSTMCell` was the naming of parameters in the `call method`. Apparently, the argument names need to match the arguments from the parent class (RNN, in this case). Therefore, I changed the name of the first argument in call from `inputs` to `sequences` to match what is expected by keras. 

<details>
<summary>Error trace>
```bash
  File "/Users/Kobi/code/sktime/test.py", line 9, in <module>
    estimator.fit(X_train,y_train)
  File "/Users/Kobi/code/sktime/sktime/regression/base.py", line 256, in fit
    self._fit(X, y)
  File "/Users/Kobi/code/sktime/sktime/regression/deep_learning/lstmfcn.py", line 177, in _fit
    self.model_ = self.build_model(self.input_shape)
  File "/Users/Kobi/code/sktime/sktime/regression/deep_learning/lstmfcn.py", line 136, in build_model
    input_layers, output_layer = self._network.build_network(input_shape, **kwargs)
  File "/Users/Kobi/code/sktime/sktime/networks/lstmfcn.py", line 88, in build_network
    x = AttentionLSTM(self.lstm_size)(x)
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/utils/traceback_utils.py", line 122, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/layers/layer.py", line 1718, in update_shapes_dict_for_target_fn
    raise ValueError(
ValueError: Exception encountered when calling AttentionLSTM.call().

For a `compute_output_shape()` method with more than one argument, all arguments should have a `_shape` suffix and match an argument from `call()`. E.g. `compute_output_shape(self, foo_shape, bar_shape)`  For layer 'AttentionLSTM', received `compute_output_shape()` argument `sequences_shape`, but `call()` does not have argument `sequences`.

Arguments received by AttentionLSTM.call():
  • args=('<KerasTensor shape=(None, 1, 24), dtype=float32, sparse=False, name=keras_tensor_1>',)
  • kwargs={'mask': 'None'}
```
</details>

3. If `AttentionLSTM` is not marked with `register_keras_serializable`, then the code fails when deserializing the model from disk(see keras [docs](https://keras.io/guides/serialization_and_saving/#custom-objects))

<details>
<summary> Error trace </summary>

```bash
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/Kobi/code/sktime/test.py", line 12, in <module>
    new_estimator=load(file_name)
  File "/Users/Kobi/code/sktime/sktime/base/_serialize.py", line 126, in load
    return cls.load_from_path(path)
  File "/Users/Kobi/code/sktime/sktime/regression/deep_learning/base.py", line 308, in load_from_path
    cls.model_ = keras.models.load_model(keras_location)
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/saving/saving_api.py", line 189, in load_model
    return saving_lib.load_model(
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/saving/saving_lib.py", line 367, in load_model
    return _load_model_from_fileobj(
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/saving/saving_lib.py", line 444, in _load_model_from_fileobj
    model = _model_from_config(
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/saving/saving_lib.py", line 433, in _model_from_config
    model = deserialize_keras_object(
  File "/Users/Kobi/code/sktime/.venv/lib/python3.10/site-packages/keras/src/saving/serialization_lib.py", line 720, in deserialize_keras_object
    raise TypeError(
TypeError: <class 'keras.src.models.functional.Functional'> could not be deserialized properly. Please ensure that components that are Python object instances (layers, models, etc.) returned by `get_config()` are explicitly deserialized in the model's `from_config()` method.
```
</details>

4. Changed `use_chrono_initialization` to `unit_forget_bias` as the former was never initialized, and, therefore, could not be loaded upon deserialization. I don't have the error trace for this one but I believe it was similar to number 3.

marcosfelt commented:
@fkiraly  Also, I realized I probably could help answer the question about breaking changes. I believe this is a breaking change due to the removal of `activity_regularizer` as an argument to `AttentionLSTMCell`.  

To prevent the breaking change, I could keep the argument and add a deprecation warning that says `activity_regularizer` will removed in a future version. What do you think?

fkiraly commented:
> What do you think?

Yes, that makes sense. The deprecation would happen in 0.37.0.

With the restriction that the network is not called directly, so we need to check how this impacts the classes calling the layer, namely `LSTMFCNClassifier` and `LSTMFCNRegressor`, e.g., via the `attention` parameter there.

marcosfelt commented:
> With the restriction that the network is not called directly, so we need to check how this impacts the classes calling the layer, namely `LSTMFCNClassifier` and `LSTMFCNRegressor`, e.g., via the `attention` parameter there.

Good point. The only time the function that contains the `AttentionLSTMCell` and `AttentionLSTM` classes is in `LSTMFCNNetwork`. And there, the only parameter used is `units` (self.lstm_size).

https://github.com/sktime/sktime/blob/99a18e61bad502f8b22cddb4e167bd36adbe2098/sktime/networks/lstmfcn.py#L87-L88

fkiraly commented:
Ok, I see - that makes sense.

May I then ask what is your use case for removal? Were you using the networks individually?

marcosfelt commented:
I was not using it personally, just trying to check off the bugs in this issue: #6153 (second point)

fkiraly reviewed (approved):
Ok, understood - thanks for the fix!

yarnabrina commented:
@fkiraly since it modifies `__init__`, should we consider it breaking and handle accordingly?

marcosfelt commented:
> @fkiraly since it modifies `__init__`, should we consider it breaking and handle accordingly?

@yarnabrina The init method is actually not publicly exposed (the class is completely defined inside another function - see https://github.com/sktime/sktime/pull/7582#issuecomment-2588474805). So I don't think we need to consider it breaking.

marcosfelt commented:
Is this PR good to merge now?
